src/t8_forest/t8_forest_ghost.cxx

```diff
@@ -1929,4 +1880,348 @@ t8_forest_ghost_destroy (t8_forest_ghost_t *pghost)
+{
+  t8_forest_ghost_interface_c *ghost_interface;
+
+  T8_ASSERT (pghost_interface != NULL);
+  ghost_interface = *pghost_interface;
+  T8_ASSERT (ghost_interface != NULL);
+
+  ghost_interface->unref ();
+}
+
+/**
+ * Abstract Base Class
+*/
+
+void
+t8_forest_ghost_interface::communicate_ownerships (t8_forest_t forest)
+{
+  t8_global_productionf (" t8_forest_ghost_interface_faces::t8_ghost_step_1_allocate \n");
+  if (forest->element_offsets == NULL) {
+    /* create element offset array if not done already */
+    memory_flag = memory_flag | CREATE_ELEMENT_ARRAY;
+    t8_forest_partition_create_offsets (forest);
+  }
+  if (forest->tree_offsets == NULL) {
+    /* Create tree offset array if not done already */
+    memory_flag = memory_flag | CREATE_TREE_ARRAY;
+    t8_forest_partition_create_tree_offsets (forest);
+  }
+  if (forest->global_first_desc == NULL) {
+    /* Create global first desc array if not done already */
+    memory_flag = memory_flag | CREATE_GFIRST_DESC_ARRAY;
+    t8_forest_partition_create_first_desc (forest);
+  }
+}
+
+void
+t8_forest_ghost_interface::communicate_ghost_elements (t8_forest_t forest)
+{
+  t8_forest_ghost_t ghost = forest->ghosts;  // TODO: make sure, that ghost is init
+  t8_ghost_mpi_send_info_t *send_info;
+  sc_MPI_Request *requests;
+
+  /* Start sending the remote elements */
+  send_info = t8_forest_ghost_send_start (forest, ghost, &requests);
+
+  /* Receive the ghost elements from the remote processes */
+  t8_forest_ghost_receive (forest, ghost);
+
+  /* End sending the remote elements */
+  t8_forest_ghost_send_end (forest, ghost, send_info, requests);
+}
+
+void
+t8_forest_ghost_interface::clean_up (t8_forest_t forest)
+{
+  t8_global_productionf (" t8_forest_ghost_interface_faces::t8_ghost_step_1_clean_up \n");
+  if (memory_flag & CREATE_GFIRST_DESC_ARRAY) {
+    /* Free the offset memory, if created */
+    t8_shmem_array_destroy (&forest->element_offsets);
+  }
+  if (memory_flag & CREATE_TREE_ARRAY) {
+    /* Free the offset memory, if created */
+    t8_shmem_array_destroy (&forest->tree_offsets);
+  }
+  if (memory_flag & CREATE_GFIRST_DESC_ARRAY) {
+    /* Free the offset memory, if created */
+    t8_shmem_array_destroy (&forest->global_first_desc);
+  }
+}
+
+/**
+ * Derived class ghost_w_search
+*/
+
+t8_forest_ghost_w_search::t8_forest_ghost_w_search (t8_ghost_type_t ghost_type): t8_forest_ghost_interface (ghost_type)
+{
+  T8_ASSERT (ghost_type != T8_GHOST_NONE);
+  T8_ASSERT (ghost_type == T8_GHOST_FACES);  // currently no other typs are suportet
+  if (ghost_type == T8_GHOST_FACES) {
+    search_fn = t8_forest_ghost_search_boundary;
+  }
+  SC_CHECK_ABORT (ghost_type != T8_GHOST_USERDEFINED,
+                  "use t8_forest_ghost_w_search(t8_forest_search_fn search_function) for userdefined ghost");
+}
+
+void
+t8_forest_ghost_w_search::do_ghost (t8_forest_t forest)
+{
+
+  communicate_ownerships (forest);
```

Comment:
Can we also put this into the if clause?

Reply:
Instead of putting all this in the `if (t8_forest_get_local_num_elements (forest) > 0)` we could start the function with
``if (t8_forest_get_local_num_elements (forest) <= 0) return;``.
So the rest of the code, does not have to be in in old if condition.